File: backend/core/src/modules/plot-points/plot-points.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { Plot, PlotPoint } from '../../entities';

@Injectable()
export class PlotPointsService {
  constructor(
    @InjectRepository(PlotPoint)
    private readonly pointsRepo: Repository<PlotPoint>,
    @InjectRepository(Plot)
    private readonly plotsRepo: Repository<Plot>,
    private readonly projectAccessService: ProjectAccessService,
  ) {}

  private async assertPlotAccess(userId: string, plotId: string): Promise<Plot> {
    const plot = await this.plotsRepo.findOne({ where: { id: plotId } });
    if (!plot) {
      throw new BadRequestException('플롯을 찾을 수 없습니다.');
    }
    await this.projectAccessService.assertProjectAccess(userId, plot.projectId);
    return plot;
  }

  async findAllForUser(userId: string, plotId: string) {
    await this.assertPlotAccess(userId, plotId);
    return this.pointsRepo.find({
      where: { plotId } as any,
      order: { orderIndex: 'ASC' } as any,
    });
  }

  async findOneForUser(userId: string, id: string) {
    const point = await this.pointsRepo.findOne({ where: { id } });
    if (!point) return null;
    await this.assertPlotAccess(userId, point.plotId);
    return point;
  }

  async createForUser(userId: string, dto: Partial<PlotPoint>) {
    if (!dto.plotId) {
      throw new BadRequestException('plotId가 필요합니다.');
    }
    await this.assertPlotAccess(userId, dto.plotId);
    const point = this.pointsRepo.create(dto as any);
    return this.pointsRepo.save(point as any);
  }

  async updateForUser(userId: string, id: string, dto: Partial<PlotPoint>) {
    const point = await this.pointsRepo.findOne({ where: { id } });
    if (!point) return null;
    await this.assertPlotAccess(userId, point.plotId);
    await this.pointsRepo.update({ id } as any, dto as any);
    return this.findOneForUser(userId, id);
  }

  async removeForUser(userId: string, id: string) {
    const point = await this.pointsRepo.findOne({ where: { id } });
    if (!point) return;
    await this.assertPlotAccess(userId, point.plotId);
    await this.pointsRepo.delete({ id } as any);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/PlotPointsService.java
package store.pjcloud.cowrite.core.service;

import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import store.pjcloud.cowrite.core.entity.Plot;
import store.pjcloud.cowrite.core.entity.PlotPoint;
import store.pjcloud.cowrite.core.repository.PlotPointRepository;
import store.pjcloud.cowrite.core.repository.PlotRepository;

@Service
public class PlotPointsService {
    private final PlotPointRepository pointsRepo;
    private final PlotRepository plotsRepo;
    private final ProjectAccessService projectAccessService;

    public PlotPointsService(PlotPointRepository pointsRepo,
                             PlotRepository plotsRepo,
                             ProjectAccessService projectAccessService) {
        this.pointsRepo = pointsRepo;
        this.plotsRepo = plotsRepo;
        this.projectAccessService = projectAccessService;
    }

    private Plot assertPlotAccess(UUID userId, UUID plotId) {
        Plot plot = plotsRepo.findById(plotId).orElse(null);
        if (plot == null) {
            throw new IllegalArgumentException("플롯을 찾을 수 없습니다.");
        }
        projectAccessService.assertProjectAccess(userId, plot.getProjectId());
        return plot;
    }

    public List<PlotPoint> findAllForUser(UUID userId, UUID plotId) {
        assertPlotAccess(userId, plotId);
        return pointsRepo.findByPlotIdOrderByOrderIndexAscCreatedAtAsc(plotId);
    }

    public PlotPoint findOneForUser(UUID userId, UUID id) {
        PlotPoint point = pointsRepo.findById(id).orElse(null);
        if (point == null) return null;
        assertPlotAccess(userId, point.getPlotId());
        return point;
    }

    @Transactional
    public PlotPoint createForUser(UUID userId, PlotPoint dto) {
        if (dto.getPlotId() == null) throw new IllegalArgumentException("plotId가 필요합니다.");
        assertPlotAccess(userId, dto.getPlotId());
        return pointsRepo.save(dto);
    }

    @Transactional
    public PlotPoint updateForUser(UUID userId, UUID id, PlotPoint dto) {
        PlotPoint existing = pointsRepo.findById(id).orElse(null);
        if (existing == null) return null;
        assertPlotAccess(userId, existing.getPlotId());
        if (dto.getTitle() != null) existing.setTitle(dto.getTitle());
        if (dto.getDescription() != null) existing.setDescription(dto.getDescription());
        if (dto.getDocumentId() != null) existing.setDocumentId(dto.getDocumentId());
        if (dto.getOrderIndex() != null) existing.setOrderIndex(dto.getOrderIndex());
        if (dto.getMetadata() != null) existing.setMetadata(dto.getMetadata());
        return pointsRepo.save(existing);
    }

    @Transactional
    public void removeForUser(UUID userId, UUID id) {
        PlotPoint existing = pointsRepo.findById(id).orElse(null);
        if (existing == null) return;
        assertPlotAccess(userId, existing.getPlotId());
        pointsRepo.deleteById(id);
    }
}
